### PR TITLE
ci: Add test for license headers

### PR DESCRIPTION
**Pull request by radeksimko**
This is a follow-up to https://github.com/hashicorp/terraform/pull/34617

In other words the aim here is to ensure we do not forget to add copyright/license headers for any future incoming code.

Here is a test run to verify this actually works: https://github.com/hashicorp/terraform/actions/runs/7799797486/job/21271331690?pr=34624#step:4:21

Screenshot 2024-02-06 at 12 40 21

## Target Release

1.8.x

### CODE REVIEW

**Comment by radeksimko** [comment]
cc'ing @radditude since I just stumbled upon [your earlier PR](https://github.com/hashicorp/terraform/pull/33702) that seems related here.

My understanding is that PR was mainly focused on generated code but did not aim to address checking of _all_ files. Correct me if I'm wrong here.

Either way, the example I linked above suggests we don't do enough checking in CI.

File: .github/workflows/checks.yml [hunk 1]
```diff
@@ -64,6 +64,17 @@ jobs:
...
           go test ./...
 
+  copywrite-tests:
+    name: "License Header Checks"
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout Repo
+        uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3
+      - name: Install copywrite
+        uses: hashicorp/setup-copywrite@v1.1.2
```
Comment (DanielMSchmidt) [inline .github/workflows/checks.yml:74]: Nit: I think we can pin the version here as well for consistency
```suggestion
        uses: hashicorp/setup-copywrite@867a1a2a064a0626db322392806428f7dc59cb3e # v1.1.2
```

**Review by DanielMSchmidt — APPROVED** [review]

**Comment by apparentlymart** [comment]
🤔 I have definitely already been called out many times by an existing check for this in our "Code Consistency Checks" job.

Does this do something different than the existing check? Should we remove the existing one in favor of this new one?

**Comment by radditude** [comment]
Huh! I was very much under the impression that our existing check added in that earlier PR would, in fact, cover all files - it uses `go generate` as the mechanism, but really it functions the same as if we just had a script running the copywrite command in the directory the file is located in. I suspect that the reason it didn't catch the missing headers you fixed in https://github.com/hashicorp/terraform/pull/34617 is just because we run the generate check _before_ the protobuf compilation check and then examine all the results together, so the copyright header added by `generate` gets wiped out by `protobuf` 😓 https://github.com/hashicorp/terraform/blob/4c0ebfff0775b66d9240050fc9ac08d87ae3be5d/.github/workflows/checks.yml#L185

I'd be totally fine with abandoning the `go generate` approach in favor of this more direct one if we think that would be easier to follow in the future! One thing to call out, though, is preserving the correct license for our various provider protocol subdirectories that have remained MPL licensed. `copywrite` doesn't really understand how to deal with that situation, so our workaround for now is separate invocations of the `copywrite` command in each directory that has a `.copywrite.hcl` config file, a la this draft from a different commit: https://github.com/hashicorp/terraform/blob/2a5ff48e3d0342c1978adf97c63414fa06c8b9ed/scripts/copyright.sh#L12-L17

**Comment by apparentlymart** [comment]
I had previously advocated for something that works with `go generate` because that's the standard way to do this sort of thing in Go and we want this codebase to be as "normal" as possible for anyone who is familiar with Go workflow, rather than having to learn a bunch of custom steps.

If that approach is making things harder then of course we can be pragmatic about it. We did already make the exception about protobuf generation because it requires a non-Go tool and only needs to be run in a specific situation that an author presumably knows applies to them (they've just edited a proto file), and I would prefer not to make more exceptions but of course it might end up being the most pragmatic thing to do, in the end. 😖

**Comment by apparentlymart** [comment]
A further thought on that subject:

We originally made the exception for protobuf because our script for that required the person running it to already have installed the right version of protoc. But in the meantime we improved the script to download the correct protoc version itself and then run it.

So I think the only remaining reason _not_ make that also be a `go generate` is that protoc is not available for all of the platforms that Go supports and that we support building Terraform on. I wonder if we could deal with that by having the protobuf script detect when it's on a platform it can't support and just print a warning and move on, and then we could make it part of `go generate` and have it do the right thing at least on the main platforms we use for development. Then we would be free of even _that_ exception, and can standardize on the normal Go workflow for everything.

Of course, that's something quite separate from the goal of this PR, so I don't mean to suggest that we ought to do that immediately here. Just thinking aloud about where we could go from here.

**Comment by radeksimko** [comment]
I reworked the changes into a make target, which should address all mentioned concerns. i.e. one can verify compliance _locally_ via `make copyright`

Thanks @radditude for the link - I only needed to make a minor change there but reused most of the code.

Test runs with a failure:

 - triggered by reverted https://github.com/hashicorp/terraform/pull/34617:
https://github.com/hashicorp/terraform/actions/runs/7874383453/job/21483868071#step:8:32
 - triggered in directories w/ MPL (which happen to be generated): https://github.com/hashicorp/terraform/actions/runs/7874717446/job/21484974347?pr=34624#step:8:76

File: Makefile [hunk 2]
```diff
@@ -26,6 +26,9 @@ staticcheck:
 exhaustive:
 	"$(CURDIR)/scripts/exhaustive.sh"
 
+copyright:
```
[the remark this answers is not in the document]
Comment (radeksimko) [reply Makefile:29]: I'm not entirely sure how easily discoverable make target arguments are so I decided to make it a new target instead - `copyrightfix`.

**Comment by radeksimko** [comment]
On the note of the two use cases (fixing vs checking) which seem to be kind of mixed together now, I wish we had a single target which does _not_ modify things and also that `fmtcheck` did what it says on the tin (i.e. _check_, not _fix_), but I will resist the attempt to fix _all the things_ in one PR. 😁

**Comment by radeksimko** [comment]
On a separate note, I have also filed the following issue upstream, with the hope that one day we may be able to get rid of this shell wrapper:

 - https://github.com/hashicorp/copywrite/issues/106

**Review by radditude — APPROVED** [review]

**Pull request merged**